src/gridcoin/upgrade.cpp

```diff
@@ -875,13 +884,123 @@ void Upgrade::DeleteSnapshot()
+                    if (filetocheck.substr(0, 3) == "blk" && filetocheck.substr(filetocheck.length() - 4, 4) == ".dat") {
+                        fs::path new_name = *Iter;
+                        new_name.replace_extension(".dat.orig");
+
+                        // Rename with orig as the extension, because ProcessBlock will load blocks into a new block data
+                        // file.
+                        fs::rename(*Iter, new_name);
+                        block_data_files.insert(new_name);
+                    }
+                }
+            }
+        }
+    } catch (fs::filesystem_error &ex) {
+        error("%s: Exception occurred: %s. Failed to rename block data files to blk*.dat.orig in preparation for "
+              "reindexing.", __func__, ex.what());
+
+        return false;
+    }
+
+    return true;
+}
+
+bool Upgrade::ReindexBlockchainData(std::set<fs::path>& block_data_files)
+{
+    bool successful = true;
+
+    try {
+        for (const auto& iter : block_data_files) {
+            FILE *block_data_file = fsbridge::fopen(iter, "rb");
+
+            LogPrintf("INFO: %s: Loading blocks from %s.", __func__, iter.filename().string());
+
+            if (!LoadExternalBlockFile(block_data_file)) {
+                successful = false;
+
+                break;
+            }
+        }
+    } catch (fs::filesystem_error &ex) {
+        error("%s: Exception occurred: %s. Failure occurred during attempt to load blocks from original "
+              "block data file(s).", __func__, ex.what());
+
+        successful = false;
+    }
+
+    if (successful) {
+        try {
+            fs::path cleanup_path;
+
+            if (!GetActualCleanupPath(cleanup_path)) return false;
+
+            fs::directory_iterator IterEnd;
+
+            // Delete the original block data files.
+            for (fs::directory_iterator Iter(cleanup_path); Iter != IterEnd; ++Iter) {
+                if (fs::is_regular_file(*Iter)) {
+                    size_t FileLoc = Iter->path().filename().string().find("blk");
+
+                    if (FileLoc != std::string::npos) {
+                        std::string filetocheck = Iter->path().filename().string();
+
+                        // Check it ends with .dat.orig and starts with blk
+                        if (filetocheck.substr(0, 3) == "blk"
+                                && filetocheck.substr(filetocheck.length() - 9, 9) == ".dat.orig") {
+                            if (!fs::remove(*Iter)) {
+                                LogPrintf("WARN: %s: Reindexing of the blockchain was successful; however, one or more of "
+                                          "the original block data files (blk*.dat.orig) was not able to be deleted. You "
+                                          "will have to delete these manually.", __func__);
+                            }
+                        }
+                    }
+                }
+            }
```

Comment:
I could probably simplify this loop, since I am passing in the set of the actual filenames, but this works fine too.

Reply:
Done.